zebra-scan/src/tests.rs

```diff
@@ -0,0 +1,371 @@
+    proto::compact_formats::{
+        self as compact, ChainMetadata, CompactBlock, CompactSaplingOutput, CompactSaplingSpend,
+        CompactTx,
+    },
+    scanning::scan_block,
+};
+use zcash_note_encryption::Domain;
+use zcash_primitives::{
+    block::BlockHash,
+    consensus::{BlockHeight, Network},
+    constants::SPENDING_KEY_GENERATOR,
+    memo::MemoBytes,
+    sapling::{
+        note_encryption::{sapling_note_encryption, SaplingDomain},
+        util::generate_random_rseed,
+        value::NoteValue,
+        Note, Nullifier, SaplingIvk,
+    },
+    zip32::{AccountId, DiversifiableFullViewingKey, ExtendedSpendingKey},
+};
+
+use color_eyre::Result;
+
+use rand::{rngs::OsRng, RngCore};
+
+use ff::{Field, PrimeField};
+use group::GroupEncoding;
+use zebra_chain::{
+    block::Block,
+    chain_tip::ChainTip,
+    serialization::{ZcashDeserializeInto, ZcashSerialize},
+    transaction::Transaction,
+};
+
+/// Prove that Zebra blocks can be scanned using the `zcash_client_backend::scanning::scan_block` function:
+/// - Populates the state with a continuous chain of mainnet blocks from genesis.
+/// - Scan the chain from the tip going backwards down to genesis.
+/// - Verifies that no relevant transaction is found in the chain when scanning for a fake account's nullifier.
+#[tokio::test]
+async fn scanning_from_populated_zebra_state() -> Result<()> {
+    let account = AccountId::from(12);
+    let vks: Vec<(&AccountId, &SaplingIvk)> = vec![];
+    let nf = Nullifier([7; 32]);
+
+    let network = zebra_chain::parameters::Network::default();
+
+    // Create a continuous chain of mainnet blocks from genesis
+    let blocks: Vec<Arc<Block>> = zebra_test::vectors::CONTINUOUS_MAINNET_BLOCKS
+        .iter()
+        .map(|(_height, block_bytes)| block_bytes.zcash_deserialize_into().unwrap())
+        .collect();
+
+    // Create a populated state service.
+    let (_state_service, read_only_state_service, latest_chain_tip, _chain_tip_change) =
+        zebra_state::populated_state(blocks.clone(), network).await;
+
+    let db = read_only_state_service.db();
+
+    // use the tip as starting height
+    let mut height = latest_chain_tip.best_tip_height().unwrap();
+
+    let mut transactions_found = 0;
+    let mut transactions_scanned = 0;
+    let mut blocks_scanned = 0;
+    while let Some(block) = db.block(height.into()) {
+        // We fake the sapling tree size to 1 because we are not in Sapling heights.
+        let sapling_tree_size = 1;
+        let orchard_tree_size = db
+            .orchard_tree_by_hash_or_height(height.into())
+            .expect("should exist")
+            .count();
+
+        let chain_metadata = ChainMetadata {
+            sapling_commitment_tree_size: sapling_tree_size
+                .try_into()
+                .expect("position should fit in u32"),
+            orchard_commitment_tree_size: orchard_tree_size
+                .try_into()
+                .expect("position should fit in u32"),
```

Comment:
Nitpick: This is part of the protocol spec, so it's not a "should", it's a requirement.

```suggestion
                .expect("sapling position is limited to u32::MAX"),
            orchard_commitment_tree_size: orchard_tree_size
                .try_into()
                .expect("orchard position is limited to u32::MAX"),
```